browser/scripts/ui/xhrForm.js

```diff
@@ -0,0 +1,288 @@
+if (typeof VizorUI === 'undefined')
+	var VizorUI = {}
+
+if (typeof uiEvent === 'undefined')
+	uiEvent = {}
+
+uiEvent.xhrFormSuccess	= 'xhrFormSuccess'		//	dispatched on document
+uiEvent.xhrFormUIReset	= 'xhrFormUIReset'		//	dispatched on form
+
+/**
+ * enables a form to submit and process its response via xhr call
+ * validation is done in backend
+ * @param $form
+ * @param onSuccess
+ */
+VizorUI.setupXHRForm = function($form, onSuccess) {	// see views/account/signup for example
+
+	var xhrEnabled = $form.data('xhrEnabled')
+	if (xhrEnabled) {
+		msg("ERROR: setupXHRForm but form already enabled " + $form.attr('id'))
+		return false
+	}
+
+	var hasFiles = $form.length && ($form[0].querySelectorAll('input[type="file"]').length > 0)
+
+	if (typeof onSuccess !== 'function') {
+		onSuccess = function(response) {
+			if (response && (typeof response.redirect !== 'undefined') && (response.redirect)) {
+				window.location.href = response.redirect
+				return
+			}
+			// append a message to form
+			var message = jQuery('<p></p>')
+							.text(response.message)
+							.addClass('xhr success message')
+			$form.append(message)
+			$form.find(':focus').trigger('blur')
+
+			var detail = {
+				id: 		$form.attr('id'),
+				response: 	response
+			}
+			var successEvent = new CustomEvent(uiEvent.xhrFormSuccess, {detail: detail});
+			document.dispatchEvent(successEvent)
+		}
+	}
+
+	$form.find('input, textarea').each(function(){
+		var $this = jQuery(this)
+		var placeholder = $this.attr('placeholder')
+		var required = $this.hasClass('required')
+		var had_error = false
+		var had_error_value = false
+		if (placeholder) {
+			if (required) placeholder += '*'
+			$this.attr('placeholder', placeholder)
+			$this.on('focus', function(){
+				had_error = $this.parent().hasClass('error')
+				had_error_value = (had_error) ? $this.val() : false
+				$this.parent().removeClass('error').find('span.message').html('')
+				$this.addClass('in_focus')
+				// ux
+				//	$this.data('placeholder', placeholder);
+				//	$this.attr('placeholder', '');
+				return true
+			});
+
+			$this.on('blur', function(){
+				// ux
+				//	$this.attr('placeholder', $this.data('placeholder'));
+				$this.removeClass('in_focus')
+				if (required && had_error &&
+					(($this.val() === '') || ($this.val() === had_error_value))
+					) $this.parent().addClass('error')
```

Comment:
a bit unreadable here, can we split the addClass from inside the if statement? almost didn't notice it